kernel/arch/riscv/paging_generic.c

```diff
@@ -63,32 +181,130 @@ map_framebuffer(pdir_t *pdir,
                 ulong size,
                 bool user_mmap)
 {
-   NOT_IMPLEMENTED();
+   if (!get_kernel_pdir())
```

Comment:
I think you said that we don't have any framebuffer support on RISCV64 yet. Also, this code mentions WC (write-combining). Do we have anything like that on RISCV ?

Reply:
Yes, there is no support for framebuffer on qemu, but many real-world chips do (such as the popular allwinner D1), I will try to port tilck to real-world chips soon, and framebuffer will be an very important feature at that time;
Some riscv chips (such as allwinner D1, which uses the T-HEAD C906 core) support WC, but it use vendor defined page attribute bits, `early_get_cpu_features()` will set `riscv_cpu_features.page_wt`(vendor defined page attribute bits) according to the specific chip model.

Reply:
Thanks for clarifying that! I'm so glad that you want to actually make Tilck to work on a real chip, not just on QEMU. That's the approach I took for x86: it must work on real machines, not just on hypervisors. And the challenge there is significant: many things that just work on QEMU don't work on real HW as-it-is due to: extra error cases, race conditions, unexpected system memory layout etc. When you get there, let me know which SoC device you decided to support first, so that I'll buy the same one. I guess it will be something similar to a RPI but for RISCV, right?

Unfortunately, at least for now Tilck cannot work on non-MMU chips, so "true embedded" microcontrollers are out of scope for the moment. But we could get there we if wanted :-)

Reply:
Has the plan to support nommu been put on the agenda?

Reply:
Are there any concrete plans to support nommu?

Reply:
@aenrbes The NOMMU support is a low-priority feature that I planned for the long term after Tilck is ported on architectures useable on embedded chips, like ARM and RISCV. Once that happened, Tilck would be much closer to the "true embedded" world. Therefore, supporting NOMMU would be mostly about the pure-software side of things:

- allow the kernel to work without any virtual memory
- prepare a special build of `busybox` using uclibc-ng which supports NOMMU

In other words, it doesn't look so hard if the kernel works already on the given architecture with MMU. Actually, about part 1, that's why we have "gcc_nocow". In that configuration `fork()` doesn't use the virtual memory and performs a full copy. Over time, we could gradually support in more and more places a NOMMU configuration where the behavior is different. 

Beyond that, after NOMMU is supported, Tilck could start having some real-time features. But all of that is very long term and low priority for now. I just wanted to hear from you what do you think about that? Does it make sense/like the idea or you would go for something completely different?

Reply:
After porting riscv64 (at least to a real-world riscv64 chip), I plan to take some time to look into the details of nuttx or linux implementations of nommu and, if possible, port those implementations to tilck too.

Reply:
I have found that there are many very interesting riscv chips (such as espressif's esp32-c3, which has a full wifi protocol stack) that don't have mmu, and I would really like tilck to run on these wireless chips (I know tilck doesn't have network support yet, But maybe lwip can be ported as a temporary solution),

Reply:
@aenrbes That would be amazing! I'm so glad to hear that you like so many aspects of this project, including the NOMMU support. 

Just a note to be on the safe side with **copyright**: I'm required to mention that, to avoid copyright violations, Tilck should not have any code close to what exists on Linux or other operating systems. It's OK to read a description about their implementation on articles etc. (e.g. on LWN), but studying closely another implementation by looking at its source code, in particular if it's from a GPL-licensed project, it's definitively a problem and we could be blamed for plagiarism. **I don't think you did that at all**. It's just a note to stay on the safe side. 

If the code is from a compatible license (MIT or BSD 2-clause), in theory we can copy it but then we need to add entry in the NOTICE file. It's still better than getting stuck, if we have no other choice. Btw, that's necessary also for `fdt_helper.c` and `fdt_helper.h` in order to be super on the safe side. 

Even if you completely re-implement the code after studying the other implementation, it could be a problem. It's legal if the other project has a permissive license, but we need to add a note saying that our implementation was heavily inspired by theirs. Because of that we still won't be able to claim "authorship of original work" for all of that code. Only with "public domain" code we're more free, but even there, it's a stretch to claim actual copyright over "public domain" code. While "public domain" comes with no conditions, the implicit condition is that we cannot claim full copyright over that and prevent others from using it. We can however re-distribute "public domain" code under any license we want. 

So.. ideally, the less we take, the better. Unless we do download components in the toolchain. That's OK for many cases because that 3rd-party source is in no way part of the project's source, which should be all result of original work. Still, even in the toolchain, I believe we cannot link GPL projects into Tilck because GPL cannot be re-distributed under a difference license, even in binary form, I think. But we don't distribute binaries for now. So, it's border-line. I hope you got the idea of the kind of problems exist with 3rd-party code.

If you need to take a quick look at some code, seriously avoid Linux and anything GPL.

Reply:
Thank you for pointing out this problem. I will pay attention to copyright things. In addition, regarding fdt_helper.c and fdt_helper.h, I will add a clarification in NOTICE file immediately.

Reply:
@aenrbes Thanks for understanding :-)